Instrument kubepug with prometheus metrics

igaskin opened the pull request:
feature to instrument kubepug with prometheus metrics. This would allow other subsystems to consume metrics for monitoring/alerting purposes.

rikatz commented:
@igaskin Thank you!!

I'll take a look this weekend together with the other PRs :D

rikatz commented on cmd/kubepug.go:
```diff
@@ -47,6 +58,11 @@ var (
+	deprecatedCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Help: "Counter of deprocated or deleted APIs",
+		Name: "deprocated_apis_count",
+	}, []string{"group", "kind", "version", "name", "scope", "object_name", "namespace", "deleted", "deprocated"})
```
s/deprocated/deprecated

rikatz commented on cmd/kubepug.go:
```diff
@@ -84,6 +103,28 @@ func runPug(cmd *cobra.Command, args []string) error {
+		signal.Notify(done, os.Interrupt)
+		go func() {
+			http.Handle("/metrics", promhttp.Handler())
+			http.ListenAndServe("0.0.0.0:8888", nil)
```
Just for security, can we make the binding address and the port also configurable?

rikatz commented on cmd/kubepug.go:
```diff
@@ -144,12 +186,17 @@ func init() {
 	rootCmd.PersistentFlags().StringVar(&format, "format", "stdout", "Format in which the list will be displayed [stdout, plain, json, yaml]")
 	rootCmd.PersistentFlags().StringVar(&filename, "filename", "", "Name of the file the results will be saved to, if empty it will display to stdout")
 	rootCmd.PersistentFlags().StringVar(&inputFile, "input-file", "", "Location of a file or directory containing k8s manifests to be analysed")
+	rootCmd.PersistentFlags().BoolVar(&monitor, "monitor", true, "run kubepug as a persistant prometheus server to monitor deprocations")
```
s/deprocations/deprecations

Also, can we make this defaulting to false to keep the original behavior of the program, and allow existing users to opt-in to the feature?

rikatz commented on lib/kubepug.go:
```diff
@@ -59,6 +66,41 @@ func (k *Kubepug) GetDeprecated() (result *results.Result, err error) {
+				"scope":       item.Scope,
+				"object_name": item.ObjectName,
+				"namespace":   item.Namespace,
+				"deprocated":  strconv.FormatBool(d.Deprecated),
```
deprocated here again :)

rikatz commented:
@igaskin thank you for the PR. It's awesome. 

I'll make some tests here to check the program behavior, and I've made some comments also :) If you can take a look into them, and please let me know when they're ready.

As soon as this is ready I'll take another look (and I promise won't take as much time as I've taken for this).

Thank you!

rikatz commented on cmd/kubepug.go:
```diff
@@ -84,6 +103,28 @@ func runPug(cmd *cobra.Command, args []string) error {
+		}()
+
+		ticker := time.NewTicker(config.ScrapeInterval)
+		for {
```
It seems that this leaves the program in an infinite loop, as an example:

```
time ./output/kubepug --monitor=false --k8s-version=v1.16.3 --scrape-interval=1m
[...]

real    3m14,432s
user    0m0,460s
sys     0m0,088s
```

I think it would be better to move the GetDeprecated to outside the cluster (as it only runs once anyway) and leave for the loop only the metrics scrapping, WDYT?

rikatz commented:
OK, so overall, also after the comments, it seems that the monitor behavior is putting the program in an infinite loop. Even putting the scrape-interval to 1m or 3s it still keeps running :)

I would move the scrapping part only to inside the for { } block, so the program is not stuck into the ticker even when I only use a kubepug --help (if you make --help only it will still keep running the program indefinitely). 

Thank you for the PR, and let me know when the proposed changes are made or if you disagree with them :)

igaskin commented:
> OK, so overall, also after the comments, it seems that the monitor behavior is putting the program in an infinite loop. Even putting the scrape-interval to 1m or 3s it still keeps running :)
> 
> I would move the scrapping part only to inside the for { } block, so the program is not stuck into the ticker even when I only use a kubepug --help (if you make --help only it will still keep running the program indefinitely).
> 
> Thank you for the PR, and let me know when the proposed changes are made or if you disagree with them :)

Ah, good point, it shouldn't try to start the monitor if its being run with the `--help` command. FWIW it was the intended behavior to put the program in an infinite loop.  The scrape-interval is period between when metrics are refreshed.  The idea here is that deprecated/deleted resources are being measured over time, and can go up or down based on changes within the cluster.

I can change it so it doesn't run indefinitely,  which in that case I think we should use a pushgateway to save the metrics.  To add some more context regarding my initial feature request, the goal here was to provide functionality to allow `kubepug` to be run persistently intra-cluster as a `deployment`.  Combined with a `service` and `servicemonitor` the cluster would become self-monitoring and allow cluster operators to create alerts to proactively respond to resources that need to be updated.

rikatz commented:
Hum, now I get it. 

So I think that:

1) without --monitor it shouldn't enter at all in the infinite loop. Imagine kubepug being used in CI, it keeps stuck in the loop even without the --monitor
2) How to deal with newer kubernetes version. IMO the monitor would need to be restarted/redeployed, right? 

About using this as a long running deployment/monitor it seems a great idea (didn't see this coming!), just wanted to keep the original behavior for existing users and then new users opt-in to the --monitor behavior :)

rikatz commented:
@igaskin just passing by, let me know when I can make another review :D

rikatz commented:
@igaskin ping :D

igaskin commented:
> @igaskin ping :D

Hey sorry I've been unresponsive on this PR.  I still plan on addressing your comments, I should have some cycles coming up this weekend to revisit.

rikatz commented:
No worries :) Just checking if this is still alive :)

igaskin commented:
closing this as stale.
